### PR TITLE
Use static instead of self where possible

### PR DESCRIPTION
**Pull request by ghost**
Ref. #17

### CODE REVIEW

**Review by foglcz — DISMISSED** [review]

**Review by OndraM — CHANGES_REQUESTED** [review]
I see few issues there:

1. Where fluent interface is used and `$this` is returned, the best is to declare `$this` as return type as well. Because declaring return type `strict` is in fact not enough - it tells its the same type as class type in current context, but `$this` tells it is also the exact same instance (what it actually is). 
2. This does not solve the original issue mentioned in #17 - extending Commands. For this phpDoc with `@return static` should be added to the named constructors (ie. wherever `return new static(...)` is used).
3. I'm not fan of using constants as "static". Its constant, so we don't expect it to be overwritten, even with LSB, no :)? Why would anyone do this :)? Or is there any usecase on your mind?

**Comment by ghost** [comment]
@OndraM 

1. should be fixed
2. ok, added it there too
3. I don't know.. does it hurt to have this option if the constants are public? But I agree that the use-cases are limited.

File: src/Matej.php [hunk 1]
```diff
@@ -32,30 +32,40 @@ public function request(): RequestBuilderFactory
...
         $this->getRequestManager()->setHttpClient($client);
 
         return $this;
     }
 
-    /** @internal used mainly in integration tests */
-    public function setBaseUrl($baseUrl): self
+    /**
+     * @internal used mainly in integration tests
+     * @return static
```
Comment (OndraM) [inline src/Matej.php]: `$this`?

File: src/Matej.php [hunk 2]
```diff
@@ -32,30 +32,40 @@ public function request(): RequestBuilderFactory
...
     {
         $this->getRequestManager()->setBaseUrl($baseUrl);
 
         return $this;
     }
 
-    /** @codeCoverageIgnore */
+    /**
+     * @codeCoverageIgnore
+     * @return static
```
Comment (OndraM) [inline src/Matej.php]: dtto, shouldn't it be `$this`?

Also one more case below

File: src/Model/Command/UserRecommendation.php [hunk 3]
```diff
@@ -78,12 +80,14 @@ public function enableHardRotation(): self
     /**
      * Define threshold of how much relevant must the recommended items be to be returned.
      * Default minimal relevance is "low".
+     *
+     * @return static
```
Comment (OndraM) [inline src/Model/Command/UserRecommendation.php]: Again `$this`? Also below

File: src/RequestBuilder/AbstractRequestBuilder.php [hunk 4]
```diff
@@ -29,6 +29,8 @@
     /**
      * If instance of RequestManager is injected to this builder object, you can build and send the request directly
      * via send() method of the builder itself.
+     *
+     * @return static
```
Comment (OndraM) [inline src/RequestBuilder/AbstractRequestBuilder.php]: $this here and below :)

**Review by OndraM — CHANGES_REQUESTED** [review]
I believe the `$this` should be used in typehints wherever fluent interface is used and `$this` is actually returned. It will still work with `$this` (instead of `static`) to cover the use case mentioned in #17 (code completion on extended classes), right?

File: src/Model/Command/UserRecommendation.php (hunk 3 above)
Comment (ghost) [reply src/Model/Command/UserRecommendation.php]: `@return static` and `@return $this` mean the same thing 🤔

Comment (OndraM) [reply src/Model/Command/UserRecommendation.php]: `static` tells its just the same *type* (resolved by LSB).
`$this` tells its not only the same type, but also the same actual instance. This means $this is "subset" of static, but not the same thing, and cannot be interchanged.

Its also recognized like this in proposed phpdoc PSR standard, see https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#keyword - points 14. and 15.

Comment (ghost) [reply src/Model/Command/UserRecommendation.php]: Cool, didn't know that. Fixed.

**Review by OndraM — APPROVED** [review]
👍 Thanks!

**Pull request merged**